src/extension/prompt/vscode-node/requestLoggerImpl.ts

```diff
@@ -24,6 +24,104 @@ import { safeStringify } from '../../../util/vs/base/common/objects';
 import { generateUuid } from '../../../util/vs/base/common/uuid';
 import { renderToolResultToStringNoBudget } from './requestLoggerToolResult';
 
+// Implementation classes with toJson methods
+class LoggedElementInfo implements ILoggedElementInfo {
+	public readonly kind = LoggedInfoKind.Element;
+
+	constructor(
+		public readonly id: string,
+		public readonly name: string,
+		public readonly tokens: number,
+		public readonly maxTokens: number,
+		public readonly trace: HTMLTracer,
+		public readonly chatRequest: any | undefined
+	) { }
+
+	toJson(): any {
+		return {
+			id: this.id,
+			kind: 'element',
+			name: this.name,
+			tokens: this.tokens,
+			maxTokens: this.maxTokens
+		};
+	}
+}
+
+class LoggedRequestInfo implements ILoggedRequestInfo {
+	public readonly kind = LoggedInfoKind.Request;
+
+	constructor(
+		public readonly id: string,
+		public readonly entry: LoggedRequest,
+		public readonly chatRequest: any | undefined
+	) { }
+
+	toJson(): any {
+		const baseInfo = {
+			id: this.id,
+			kind: 'request',
+			type: this.entry.type,
+			name: this.entry.debugName
+		};
+
+		if (this.entry.type === LoggedRequestKind.MarkdownContentRequest) {
+			return {
+				...baseInfo,
+				startTime: new Date(this.entry.startTimeMs).toISOString()
+			};
+		} else {
+			return {
+				...baseInfo,
+				startTime: this.entry.startTime?.toISOString(),
+				endTime: this.entry.endTime?.toISOString(),
+				duration: this.entry.endTime && this.entry.startTime ?
+					this.entry.endTime.getTime() - this.entry.startTime.getTime() : undefined,
+				model: 'model' in this.entry.chatParams ? this.entry.chatParams.model : undefined,
+				messages: 'messages' in this.entry.chatParams ? this.entry.chatParams.messages : undefined,
+				usage: this.entry.type === LoggedRequestKind.ChatMLSuccess ? this.entry.usage : undefined,
+				timeToFirstToken: this.entry.type !== LoggedRequestKind.ChatMLCancelation ? this.entry.timeToFirstToken : undefined,
+				result: this.entry.type !== LoggedRequestKind.ChatMLCancelation ? this.entry.result : undefined,
+			};
+		}
+	}
+}
+
+class LoggedToolCall implements ILoggedToolCall {
+	public readonly kind = LoggedInfoKind.ToolCall;
+
+	constructor(
+		public readonly id: string,
+		public readonly name: string,
+		public readonly args: unknown,
+		public readonly response: LanguageModelToolResult2,
+		public readonly chatRequest: any | undefined,
+		public readonly time: number,
+		public readonly thinking?: ThinkingData
+	) { }
+
+	async toJson(): Promise<any> {
+		const result: string[] = [];
+		for (const content of this.response.content as (LanguageModelTextPart | LanguageModelPromptTsxPart)[]) {
```

Comment:
Tools can also return `LanguageModelDataPart` which we should handle here (e.g. images, most commonly returned by MCP tools today)